### PR TITLE
feature: batch user deletion support for mixpanel and clevertap

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change

- as of now for existing destination if we want to delete users then we are doing it one by one. like for every user we are making user deletion request to destination which is impecting the performance of the api
- in this pr for mixpanel and clevertap destination we are making user deletion request in batch since they are supporting batch user deletion
- also, for a perticular user if we don't found required key(ex **userId**) then we will drop that perticular user and proceed for remaining users

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ✅] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ✅] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/mp/deleteUsers.js [hunk 1]
```diff
@@ -20,33 +20,29 @@ const userDeletionHandler = async (userAttributes, config) => {
...
-        .setMessage("Could not get response")
-        .setStatus(500)
-        .build();
-    }
+  const response = await httpPOST(endpoint, data, headers);
+  if (!response || !response.response) {
+    throw new ErrorBuilder()
+      .setMessage("Could not get response")
+      .setStatus(500)
+      .build();
```
Comment (utsabc) [inline v0/destinations/mp/deleteUsers.js:45]: We need to handle response properly you
```
const deletionRespones = await httpPOST(endpoint, data, headers);
 const processedDeletionRespones = processAxiosResponse(deletionRespones);
 if ( processedDeletionRespones.status !== 200) {
     throw new ErrorBuilder()
      .setMessage("[Mixpanel]::Deletion Request is not successful")
      .setStatus(processedDeletionRespones.status)
      .build();
 }

File: v0/destinations/clevertap/deleteUsers.js [hunk 2]
```diff
@@ -30,23 +30,19 @@ const userDeletionHandler = async (userAttributes, config) => {
...
+    if (userAttribute.userId) {
+      identity.push(userAttribute.userId);
     }
+  });
+  const response = await httpPOST(endpoint, { identity }, { headers });
+  if (!response || !response.response) {
+    throw new ErrorBuilder()
+      .setMessage("Could not get response")
+      .setStatus(500)
+      .build();
```
Comment (utsabc) [inline v0/destinations/clevertap/deleteUsers.js:45]: Need to add correct handling same as below

File: v0/destinations/af/deleteUsers.js [hunk 3]
```diff
@@ -42,116 +126,66 @@ const userDeletionHandler = async (userAttributes, config) => {
...
+        .build();
+    }
+    const response = await processUserDeletionRequest(
+      iosUsers,
+      endpoint,
+      config.appleAppId,
+      statusCallbackUrls
     );
-    if (filteredStatusCallbackUrlsArray.length > 3) {
+    if (!response || !response.response) {
```
Comment (utsabc) [inline v0/destinations/af/deleteUsers.js:153]: We need to update the response handling here as well

File: v0/destinations/clevertap/deleteUsers.js [hunk 4]
```diff
@@ -30,23 +33,20 @@ const userDeletionHandler = async (userAttributes, config) => {
...
-      throw new ErrorBuilder()
-        .setMessage("Could not get response")
-        .setStatus(500)
-        .build();
+  const identity = [];
+  userAttributes.forEach(userAttribute => {
+    // Dropping the user if userId is not present
+    if (userAttribute.userId) {
+      identity.push(userAttribute.userId);
     }
```
Comment (ItsSudip) [inline v0/destinations/clevertap/deleteUsers.js:41]: Can we verify how many users we can delete in a single batch? I mean is there any limitation in the API in sending userId for deletion?

File: v0/destinations/mp/deleteUsers.js [hunk 5]
```diff
@@ -20,33 +23,30 @@ const userDeletionHandler = async (userAttributes, config) => {
       .setStatus(400)
       .build();
   }
-  const params = {
-    data: {
-      $token: `${config.token}`,
-      $delete: null,
-      $ignore_alias: true
+  const endpoint = "https://api.mixpanel.com/engage#profile-delete";
```
Comment (ItsSudip) [inline v0/destinations/mp/deleteUsers.js:26]: We need to set the endpoint depending on config.dataResidency.

File: v0/destinations/mp/deleteUsers.js [hunk 6]
```diff
@@ -20,33 +23,30 @@ const userDeletionHandler = async (userAttributes, config) => {
...
+  const data = [];
+  userAttributes.forEach(userAttribute => {
+    // Dropping the user if userId is not present
+    if (userAttribute.userId) {
+      data.push({
+        $token: `${config.token}`,
+        $distinct_id: userAttribute.userId,
+        $delete: null,
+        $ignore_alias: true
```
Comment (ItsSudip) [inline v0/destinations/mp/deleteUsers.js:35]: Apart from userId every attribute is the same for all. We can store them in a variable and use, WDYT.

File: v0/destinations/mp/deleteUsers.js [hunk 7]
```diff
@@ -20,33 +23,30 @@ const userDeletionHandler = async (userAttributes, config) => {
...
-    params.data.$distinct_id = userAttributes[i].userId;
-    // eslint-disable-next-line no-await-in-loop
-    const response = await httpPOST(endpoint, null, { params });
-    if (!response || !response.response) {
-      throw new ErrorBuilder()
-        .setMessage("Could not get response")
-        .setStatus(500)
-        .build();
-    }
+  const deletionRespone = await httpPOST(endpoint, data, headers);
```
Comment (ItsSudip) [inline v0/destinations/mp/deleteUsers.js:43]: Here also we should verify the limit of API.

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: v0/destinations/mp/deleteUsers.js [hunk 8]
```diff
@@ -20,33 +23,30 @@ const userDeletionHandler = async (userAttributes, config) => {
...
+  const processedDeletionRespone = processAxiosResponse(deletionRespone);
+  if (processedDeletionRespone.status !== 200) {
```
Comment (utsabc) [inline v0/destinations/mp/deleteUsers.js:45]:
```suggestion
  if (!isHttpStatusSuccess(processedDeletionRespone.status)) {
```
Use this function as success can be in range of 200 to 300

File: v0/destinations/clevertap/deleteUsers.js [hunk 9]
```diff
@@ -30,23 +33,20 @@ const userDeletionHandler = async (userAttributes, config) => {
...
+  });
+  const deletionRespone = await httpPOST(endpoint, { identity }, { headers });
+  const processedDeletionRespone = processAxiosResponse(deletionRespone);
+  if (processedDeletionRespone.status !== 200) {
```
Comment (utsabc) [inline v0/destinations/clevertap/deleteUsers.js:45]: Same here

**Review by utsabc — DISMISSED** [review]

**Review by ItsSudip — APPROVED** [review]

**Review by anantjain45823 — APPROVED** [review]

File: __tests__/data/clevertap_deleteUsers_proxy_input.json [hunk 10]
```diff
@@ -42,28 +42,5 @@
         }
       ]
     }
```
Comment (krishna2020) [inline __tests__/data/clevertap_deleteUsers_proxy_input.json:44]: we need to come up with unit test cases here to validate the batching logic

**Pull request merged**